### DISCUSSION TITLE
Send multiple objects

### DISCUSSION THREAD

**Discussion by florian-gandon** [finos/FDC3#979] [Q&A] [answered]
Hello !

Do we have described how we should manage sending multiple type of objects ?

One after another ? For instance, if I send a country and then an instrument. Should we reset or remove fields or id no present in the instument ?

Together ?Like for instance:
data {
  instrument {
    type: "fdc3.instrument",
    name: "Microsoft",
    id: {
        ISIN: "US5949181045"
    }
  }
  country {
    type: "fdc3.country",
    name: "Sweden"
  }
}

I've found a discussion talking about that, https://github.com/finos/FDC3/discussions/545, but nothing concrete and I'm wondering if it was resolved.

Thank you !
Florian Gandon

**Comment by kriswest** [comment, marked as answer]
Hi Florian,

Please see
https://fdc3.finos.org/docs/api/spec#broadcasting-and-listening-for-multiple-context-types
for the current advice on this. It refers to complex types that are
composed of other types and recommends broadcasting the simpler subtypes
first, followed by the complex type in quick succession. The same applies
to multiple types.

K

**Reply by florian-gandon** [reply]
Thank you for the page ! It clarifies some points. I should have read it until the end.
One last point, I guess we should actually reset (opposite to merge) other properties like the "id" when sending a country after a instrument. Am I right ?

**Reply by kriswest** [reply]
I didn't fully understand the follow-up question, but I'll have a go at answering anyway:

Different context objects should be considered separately. Hence, the `id` property (and its sub-fields) of an `fdc3.country` relates to the specific country, while `id` in an instrument relates to the security or index it represents. Nothing _should_ be inferred from them arriving one after another (after all, they could actually come from different sources) nor should their construction be affected by each other. Hence, I _think_ I agree with the above statement (don't merge ids across the two different types). 

If you need to express some meaning through the combination of a country and an instrument then you would need to create a new context that contained both in its own fields (embedding the whole country and instrument contexts). You can then broadcast all 3 types, with other apps receiving or processing only the ones they support (e.g. a pricing chart might listen for instrument, some visualization the country and something else the complex type containing both). 

Does that help clarify?

**Reply by florian-gandon** [reply]
It helped.
I was wondering how manage one object coming after another and you're right, bascially one should not be affected or interfer by another (whatever the order of arriving).

Thank you !

**Reply by kriswest** [reply]
No problem. Worth noting there are two ways to go with context listeners: you can either have one for all types of context (pass `null` as the contextType argument to `addContextListener`, us an `if`/`else` or `switch` statement etc. to process each) or one per type (pass the type as the first argument and then supply a handler specific to the type). In either case,  if you handle multiple types in your app, assume the simplest arrives first (as per the advice on order of broadcase) and the more complex after, so that the more complex is last/can supersede the reaction to the simpler types if necessary.

Whilst this wasn't my preferred way to do this (when we wrote the advice at https://fdc3.finos.org/docs/api/spec#broadcasting-and-listening-for-multiple-context-types, it should be workable. Do let us know how you get on with it!
